### PR TITLE
Update intro.rst to QGIS 3 (PyQGIS cookbook)

### PR DESCRIPTION
**Pull request by havatv**
Updated to QGIS 3.
The QGIS console screenshot has not yet been updated (but the commands shown still work in QGIS 3).

### CODE REVIEW

**Comment by havatv** [comment]
I have now also updated console.png (to QGIS 3).

File: source/docs/pyqgis_developer_cookbook/intro.rst [hunk 1]
```diff
@@ -107,14 +110,16 @@ The console opens as a non-modal utility window:
 
 The screenshot above illustrates how to get the layer currently selected
 in the layer list, show its ID and optionally, if it is a vector layer,
-show the feature count. For interaction with QGIS environment, there is a
+show the feature count (the screenshot is of QGIS 2, but the same
+commands work in QGIS 3).
```
Comment (ghtmtt) [inline source/docs/pyqgis_developer_cookbook/intro.rst]: screenshot is updated (thanks!) so also the text should be updated (~~QGIS 2~~)

File: source/docs/pyqgis_developer_cookbook/intro.rst [hunk 2]
```diff
@@ -131,8 +136,9 @@ Python Plugins
 ==============
 
 QGIS allows enhancement of its functionality using plugins. This
-was originally possible only with C++ language. With the addition of Python
-support to QGIS, it is also possible to use plugins written in Python.
+was originally possible only using the C++ language.
+With the addition of Python
```
Comment (ghtmtt) [inline source/docs/pyqgis_developer_cookbook/intro.rst]: weird line wraps..

Comment (havatv) [reply source/docs/pyqgis_developer_cookbook/intro.rst]: It was just to limit the changes.  I will fix it.

File: source/docs/pyqgis_developer_cookbook/intro.rst (hunk 1 above)
Comment (havatv) [reply source/docs/pyqgis_developer_cookbook/intro.rst]: Sorry, forgot about that.  Will fix.

**Comment by DelazJ** [comment]
Just to mention that @anitagraser began a refactoring of this chapter. If you want to share your thoughts or pick some parts of her work it's at #2541

**Comment by havatv** [comment]
I have fixed the issues in the last commit.

**Comment by havatv** [comment]
I saw the pull request by @anitagraser after I had started editing.  I was hoping that her suggested "structure" changes can be fixed later.

File: source/docs/pyqgis_developer_cookbook/intro.rst [hunk 3]
```diff
@@ -12,68 +12,73 @@ Introduction
...
 * automatically run Python code when QGIS starts
 * issue commands in Python console within QGIS
 * create and use plugins in Python
-* create custom applications based on QGIS API
+* create custom applications based on the QGIS API
 
 
 Python bindings are also available for QGIS Server:
 
 * starting from 2.8 release, Python plugins are also available on QGIS Server
```
Comment (DelazJ) [inline source/docs/pyqgis_developer_cookbook/intro.rst]: I'd remove `starting from 2.8 release,`

File: source/docs/pyqgis_developer_cookbook/intro.rst [hunk 4]
```diff
@@ -12,68 +12,73 @@ Introduction
...
   (see :ref:`Server Python Plugins <server_plugins>`)
-* starting from 2.11 version (Master at 2015-08-11), QGIS Server library has
-  Python bindings that can be used to embed QGIS Server into a Python application.
+* starting from 2.11 version (Master at 2015-08-11), QGIS Server
```
Comment (DelazJ) [inline source/docs/pyqgis_developer_cookbook/intro.rst]: ditto about version  mention

File: source/docs/pyqgis_developer_cookbook/intro.rst [hunk 5]
```diff
@@ -12,68 +12,73 @@ Introduction
...
+  into a Python application.
 
 .. index:: API
 
-There is a `complete QGIS API <http://qgis.org/api/>`_ reference that documents
-the classes from the QGIS libraries. Pythonic QGIS API is nearly identical
-to the API in C++.
+There is a `complete QGIS API <http://qgis.org/api/>`_ reference that
+documents the classes from the QGIS libraries. Pythonic QGIS API is
+nearly identical to the API in C++.
```
Comment (DelazJ) [inline source/docs/pyqgis_developer_cookbook/intro.rst]: we should mention the pyqgis site here, shouldn't we? (maybe in replacement of the c++ one)

File: source/docs/pyqgis_developer_cookbook/intro.rst [hunk 6]
```diff
@@ -95,60 +103,64 @@ it is executed by the embedded Python interpreter.
...
-Many plugins covering various functionality have been written since the
-introduction of Python support. The plugin installer allows users to easily
-fetch, upgrade and remove Python plugins. See the `Python Plugin Repositories
-<http://www.qgis.org/wiki/Python_Plugin_Repositories>`_ page for various
-sources of plugins.
+Many plugins covering various functionality have been written since
+the introduction of Python support. The plugin installer allows users
+to easily fetch, upgrade and remove Python plugins.
+See the `Python Plugin Repositories
+<http://www.qgis.org/wiki/Python_Plugin_Repositories>`_ page for
```
Comment (DelazJ) [inline source/docs/pyqgis_developer_cookbook/intro.rst]: This link is stale; i guess it should be something like https://plugins.qgis.org/plugins/

File: source/docs/pyqgis_developer_cookbook/intro.rst [hunk 7]
```diff
@@ -160,35 +172,38 @@ instructions.
...
+locate the QGIS resources such as projection information, providers
+for reading vector and raster layers, etc. QGIS Resources are
+initialized by adding a few lines to the beginning of your application
+or script. The code to initialize QGIS for custom applications and
+standalone scripts is similar, but examples of each are provided
+below.
 
-Note: do *not* use :file:`qgis.py` as a name for your test script --- Python
-will not be able to import the bindings as the script's name will shadow them.
+Note: do *not* use :file:`qgis.py` as a name for your test script ---
```
Comment (DelazJ) [inline source/docs/pyqgis_developer_cookbook/intro.rst]: could we use .. note:: here (more visible)

File: source/docs/pyqgis_developer_cookbook/intro.rst [hunk 8]
```diff
@@ -283,30 +306,32 @@ QGIS installation path:
 * on Linux: :command:`export PYTHONPATH=/<qgispath>/share/qgis/python`
 * on Windows: :command:`set PYTHONPATH=c:\\<qgispath>\\python`
```
Comment (DelazJ) [inline source/docs/pyqgis_developer_cookbook/intro.rst]: any macOS user to tell what's the corresponding link here, please? @Gustry ?

**Comment by DelazJ** [comment]
> I saw the pull request by @anitagraser after I had started editing. I was hoping that her suggested "structure" changes can be fixed later.

Sure, no special request/requirement in my message. The restructuring which is also about finding the place to move some description to lighten this introduction can be done later. My idea was mainly to inform you and if possible, "pick" the contents/ideas so that this part of the refactoring would be done.

File: source/docs/pyqgis_developer_cookbook/intro.rst (hunk 8 above)
Comment (Gustry) [reply source/docs/pyqgis_developer_cookbook/intro.rst]: I'm not on MacOS anymore (back to Ubuntu). I guess one of my previous colleagues will tell you @timlinux @ismailsunni ?

File: source/docs/pyqgis_developer_cookbook/intro.rst [hunk 9]
```diff
@@ -82,9 +87,12 @@ homebrew or MacPorts installs on Mac.
 The :file:`startup.py` file
 ----------------------------
 
-Every time QGIS starts, the user's Python home directory (usually:
-:file:`.qgis2/python`) is searched for a file named :file:`startup.py`, if that file exists,
-it is executed by the embedded Python interpreter.
+Every time QGIS starts, the user's Python home directory (Linux:
+:file:`.local/share/QGIS/QGIS3/profiles/default/python`, Windows:
+:file:`AppData\Roaming\QGIS\QGIS3\profiles\default/python`, Mac OS X:
```
[the remark this answers is not in the document]
Comment (havatv) [reply source/docs/pyqgis_developer_cookbook/intro.rst]: Done in the latest commit.

File: source/docs/pyqgis_developer_cookbook/intro.rst [hunk 10]
```diff
@@ -82,9 +87,12 @@ homebrew or MacPorts installs on Mac.
...
+:file:`Library/Application Support/QGIS/QGIS3/profiles/default`) is
```
[the remark this answers is not in the document]
Comment (havatv) [reply source/docs/pyqgis_developer_cookbook/intro.rst]: Done in the latest commit.

File: source/docs/pyqgis_developer_cookbook/intro.rst [hunk 11]
```diff
@@ -12,68 +12,73 @@ Introduction
...
+Python language. We've decided for Python as it's one of the most
+favourite languages for scripting. PyQGIS bindings depend on SIP and
+PyQt. The reason for using SIP instead of more widely used SWIG is
+that the whole QGIS code depends on Qt libraries. Python bindings for
+Qt (PyQt) are done also using SIP and this allows seamless integration
+of PyQGIS with PyQt.
 
-There are several ways how to use Python bindings in QGIS desktop, they are covered
-in detail in the following sections:
+There are several ways how to use Python bindings in QGIS desktop,
```
[the remark this answers is not in the document]
Comment (havatv) [reply source/docs/pyqgis_developer_cookbook/intro.rst]: Done in the latest commit.

File: source/docs/pyqgis_developer_cookbook/intro.rst (hunk 3 above)
Comment (havatv) [reply source/docs/pyqgis_developer_cookbook/intro.rst]: Done in the latest commit.

File: source/docs/pyqgis_developer_cookbook/intro.rst (hunk 4 above)
Comment (havatv) [reply source/docs/pyqgis_developer_cookbook/intro.rst]: Done in the latest commit.

File: source/docs/pyqgis_developer_cookbook/intro.rst (hunk 5 above)
Comment (havatv) [reply source/docs/pyqgis_developer_cookbook/intro.rst]: Done in the latest commit.

File: source/docs/pyqgis_developer_cookbook/intro.rst (hunk 6 above)
Comment (havatv) [reply source/docs/pyqgis_developer_cookbook/intro.rst]: Done in the latest commit.

File: source/docs/pyqgis_developer_cookbook/intro.rst (hunk 7 above)
Comment (havatv) [reply source/docs/pyqgis_developer_cookbook/intro.rst]: Done in the latest commit.

**Comment by havatv** [comment]
It would be good if this pull request could be merged, so that others (that knows more about this) can take over and improve.

File: source/docs/pyqgis_developer_cookbook/intro.rst (hunk 1 above)
Comment (havatv) [reply source/docs/pyqgis_developer_cookbook/intro.rst]: The text has been fixed (QGIS 2 removed).

File: source/docs/pyqgis_developer_cookbook/intro.rst (hunk 2 above)
Comment (havatv) [reply source/docs/pyqgis_developer_cookbook/intro.rst]: Has been fixed.

File: source/docs/pyqgis_developer_cookbook/intro.rst (hunk 8 above)
Comment (ismailsunni) [reply source/docs/pyqgis_developer_cookbook/intro.rst]: I add this path: `/Applications/QGIS.app/Contents/Resources/python` to my python interpreter in my PyCharm.

Comment (havatv) [reply source/docs/pyqgis_developer_cookbook/intro.rst]: I added this below the list for linux and winows in the latest commit.

**Pull request merged**